pkg/odo/genericclioptions/runnable.go

```diff
@@ -26,9 +44,28 @@ func GenericRun(o Runnable, cmd *cobra.Command, args []string) {
 	// LogErrorAndExit is used so that we get -o (jsonoutput) for cmds which have json output implemented
 	util.LogErrorAndExit(checkConflictingFlags(cmd), "")
 	// Run completion, validation and run.
-	util.LogErrorAndExit(o.Complete(cmd.Name(), cmd, args), "")
-	util.LogErrorAndExit(o.Validate(), "")
-	util.LogErrorAndExit(o.Run(), "")
+	// Only upload data to segment for completion and validation if a non-nil error is returned.
+	err = o.Complete(cmd.Name(), cmd, args)
+	if err != nil {
+		uploadToSegmentAndLog(cmd, err, startTime)
+	}
+	err = o.Validate()
+	if err != nil {
+		uploadToSegmentAndLog(cmd, err, startTime)
+	}
+	uploadToSegmentAndLog(cmd, o.Run(), startTime)
+}
+
+// uploadToSegmentAndLog uploads the data to segment
+func uploadToSegmentAndLog(cmd *cobra.Command, err error, startTime time.Time) {
+	if serr := segmentClient.Upload(cmd.CommandPath(), time.Since(startTime), err); serr != nil {
```

Comment:
is this async in nature?

Reply:
As far as I can tell, yes. The way it works is, calling the `Upload` method adds data to the queue and BTS sends it asynchronously in a batch once the connection closes.

Reply:
Can we confirm that?

Reply:
> Can we confirm that?

It works as @valaparthvi describes